Add cachebench

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #147546
* __->__ #147537

This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

oulgen committed:
Add cachebench

[ghstack-poisoned]

jamesjwu commented on benchmarks/dynamo/cachebench.py:
```diff
@@ -0,0 +1,173 @@
+        for dynamic in [False, True]:
+            args = list(base_args)
+            if mode == "inference":
+                args.extend(["--inference", "--bfloat16"])
```
Is bfloat16 gonna always work for every model?

jamesjwu replied:
Similar question for --amp

jamesjwu commented on benchmarks/dynamo/cachebench.py:
```diff
@@ -0,0 +1,173 @@
+    base_args = [
+        sys.executable,
+        torchbench_file,
+        f"--only={model}",
```
I wonder if there's a way to make torchbench take a list of models instead so we can just run them all, and then read the resulting JSON once

oulgen replied:
discussed offline, we cannot do this since I dont want in memory cache sharing

jamesjwu commented on benchmarks/dynamo/cachebench.py:
```diff
@@ -0,0 +1,173 @@
+def _write_results_to_json(results: list[RunResult], output_filename: str) -> None:
+    records = []
+    for result in results:
+        for metric_name in ["cold_compile_time(s)", "warm_compile_time(s)", "speedup"]:
```
nit: These if statements seem kinda silly, just do for (metric_name, value) in [("cold_compile_time", result.cold_compile_s), ()] 
etc
You can use a constant dict instead of a list of tuples too

jamesjwu commented on benchmarks/dynamo/cachebench.py:
```diff
@@ -0,0 +1,173 @@
+    for mode in ["inference", "training"]:
+        for dynamic in [False, True]:
+            args = list(base_args)
+            if mode == "inference":
```
just do args.extend(f"--{mode}")

jamesjwu commented on benchmarks/dynamo/cachebench.py:
```diff
@@ -0,0 +1,173 @@
+    with fresh_inductor_cache():
+        env = os.environ.copy()
+        with tempfile.NamedTemporaryFile(suffix=".csv") as file:
+            args.append("--output=" + file.name)
```
Doesn't this thing return a JSON as well? Can you parse the JSON instead?

oulgen replied:
nope this is csv, there's also a json but that's huge, so i think this is simpler

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen committed:
None on "Add cachebench"


This PR adds a new benchmark called cachebench in order to measure/demonstrate the prowess of PT2 caching.
```
python benchmarks/dynamo/cachebench.py --output="result.json"
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge
